types/wlr_seat.c

```diff
@@ -819,3 +889,259 @@ void wlr_seat_keyboard_notify_key(struct wlr_seat *seat, uint32_t time,
+
+	if (grab != wlr_seat->touch_state.default_grab) {
+		wlr_seat->touch_state.grab = wlr_seat->touch_state.default_grab;
+		wl_signal_emit(&wlr_seat->events.touch_grab_end, grab);
+		if (grab->interface->cancel) {
+			grab->interface->cancel(grab);
+		}
+	}
+}
+
+static void touch_point_destroy(struct wlr_touch_point *point) {
+	wl_signal_emit(&point->events.destroy, point);
+	wl_list_remove(&point->surface_destroy.link);
+	wl_list_remove(&point->resource_destroy.link);
+	wl_list_remove(&point->link);
+	free(point);
+}
+static void handle_touch_point_resource_destroy(struct wl_listener *listener,
+		void *data) {
+	struct wlr_touch_point *point =
+		wl_container_of(listener, point, resource_destroy);
+	touch_point_destroy(point);
+}
+
+static void handle_touch_point_surface_destroy(struct wl_listener *listener,
+		void *data) {
+	struct wlr_touch_point *point =
+		wl_container_of(listener, point, surface_destroy);
+	touch_point_destroy(point);
+}
+
+static struct wlr_touch_point *touch_point_create(
+		struct wlr_seat *seat, int32_t touch_id,
+		struct wlr_surface *surface, double sx, double sy) {
+	struct wl_client *wl_client = wl_resource_get_client(surface->resource);
+	struct wlr_seat_client *client = wlr_seat_client_for_wl_client(seat, wl_client);
+
+	if (!client || !client->touch) {
+		// touch points are not valid without a connected client with touch
+		return NULL;
+	}
+
+	struct wlr_touch_point *point = calloc(1, sizeof(struct wlr_touch_point));
+	if (!point) {
+		return NULL;
+	}
+
+	point->touch_id = touch_id;
+	point->surface = surface;
+	point->client = client;
+
+	point->sx = sx;
+	point->sy = sy;
+
+	wl_signal_init(&point->events.destroy);
+
+	wl_signal_add(&surface->events.destroy, &point->surface_destroy);
+	point->surface_destroy.notify = handle_touch_point_surface_destroy;
+	wl_resource_add_destroy_listener(surface->resource,
+		&point->resource_destroy);
+	point->resource_destroy.notify = handle_touch_point_resource_destroy;
+
+	wl_list_insert(&seat->touch_state.touch_points, &point->link);
+
+	return point;
+}
+
+struct wlr_touch_point *wlr_seat_touch_get_point(
+		struct wlr_seat *seat, int32_t touch_id) {
+	struct wlr_touch_point *point = NULL;
+	wl_list_for_each(point, &seat->touch_state.touch_points, link) {
+		if (point->touch_id == touch_id) {
+			return point;
+		}
+	}
+
+	return NULL;
+}
+
+void wlr_seat_touch_notify_down(struct wlr_seat *seat,
+		struct wlr_surface *surface, uint32_t time, int32_t touch_id, double sx,
+		double sy) {
+	clock_gettime(CLOCK_MONOTONIC, &seat->last_event);
+	struct wlr_seat_touch_grab *grab = seat->touch_state.grab;
+	struct wlr_touch_point *point =
+		touch_point_create(seat, touch_id, surface, sx, sy);
+	if (!point) {
+		wlr_log(L_ERROR, "could not create touch point");
+		return;
+	}
+
+	grab->interface->down(grab, time, point);
+
+	if (wl_list_length(&seat->touch_state.touch_points) == 1) {
+		seat->touch_state.grab_serial = wl_display_get_serial(seat->display);
```

Comment:
I've seen you're using `wl_display_get_serial` like this at several places. Is it really okay to do this? I'd prefer `grab->interface->down` to return the serial for instance.

Reply:
Yeah that's probably better.